### PR TITLE
[Hinty] Type hinting automaton.py

### PR DESCRIPTION
**Pull request by Davidy22**
Used ansmachine.py as initial reference, Looked at some other files that I grepped into along the way.

Some notes taken while adding the hints:
add_breakpoints() returns None
Unsure of object type that could have fileno() on 609
645 assumed from rd and wr from the above class

Part of the long series of commits to work towards completion of #2158

Tests fail on my branch, but the error message seems unrelated to the file I changed, and I tried committing just a copy of master and that fails too so I guess it's just some teething issue in master

### CODE REVIEW

**Comment by Davidy22** [comment]
How did the % on files that I didn't change get affected?

**Comment by polybassa** [comment]
Don't care about it. Codecov has some instability.

**Comment by gpotter2** [comment]
You need to add `automaton.py` to `.config/mypy/mypy_enabled.txt`

**Comment by Davidy22** [comment]
Added, noticed flake had some things to say about my hints too.

**Comment by gpotter2** [comment]
Thats some very solid work thanks a lot !

I have a trick for the `WINDOWS` vs `sys.platform` issue in the bags, because we're going to need it at some point to type arch/ anyways. I'mma try to commit it ASAP

File: test/contrib/automotive/ccp.uts [hunk 1]
```diff
@@ -884,9 +884,8 @@ sock2 = TestSocket(CAN)
 sock1.pair(sock2)
 
 def answer(pkt):
-    cro = CRO(pkt.data)
```
Comment (gpotter2) [inline test/contrib/automotive/ccp.uts:887]: @polybassa I had to change this, I don't understand what this is.
You are sending a `CCP(identifier=0x700)/CRO(ctr=0x53)/PROGRAM_6(data=b"\x10\x11\x12\x10\x11\x12")`, extract `pkt.data` then try to parse that to a `CRO`.. Is this supposed to work?

File: test/testsocket.py [hunk 2]
```diff
@@ -21,7 +21,7 @@
 open_test_sockets = list()  # type: List[TestSocket]
 
 
-class TestSocket(ObjectPipe, object):
+class TestSocket(ObjectPipe[Packet], SuperSocket):
```
Comment (polydroi) [inline test/testsocket.py:24]: I think I've tested this, but than decided to implement the Supersocket functions directly, since I had problems with the close function. Maybe that's fixed now, I just wanted to mention this, in case of strange bugs.

Comment (gpotter2) [reply test/testsocket.py:24]: This was indeed super buggy but I think I figured it out, it needed an overwrite of the `recv()` function

Comment (gpotter2) [reply test/testsocket.py:24]: Sorry for the previous ping btw 😅

**Comment by polybassa** [comment]
Your welcome. I was searching for the question, but couldn't find it. 
-- 
Viele Grüße

Nils Weiß

On November 1, 2021 10:48:16 PM GMT+01:00, gpotter2 ***@***.***> wrote:
***@***.*** commented on this pull request.>

**Comment by polybassa** [comment]
Also, select or select_objects made a problem. But again, maybe it's all fixed now. Just to keep in mind. 
-- 
Viele Grüße

Nils Weiß

On November 1, 2021 10:47:54 PM GMT+01:00, gpotter2 ***@***.***> wrote:
***@***.*** commented on this pull request.

**Comment by gpotter2** [comment]
@guedou This is ready for review.
I'd like to finish typing all of scapy's core soon-ish and proceed with 2.5.0.

There is mostly pipetools and arch left

File: test/contrib/automotive/ccp.uts [hunk 3]
```diff
@@ -879,6 +879,7 @@ assert dto.hashret() == cro.hashret()
 
 = CAN Socket sr1 with dto.ansers(cro) == True
 
+conf.debug_dissector = True
```
Comment (guedou) [inline test/contrib/automotive/ccp.uts:882]: Is this change needed?

File: scapy/automaton.py [hunk 4]
```diff
@@ -31,11 +31,32 @@
...
+    Type,
+    TypeVar,
+    Union,
+    _Generic_metaclass,
+    cast,
+)
+
 
 def select_objects(inputs, remain):
+    # type: (List[Any], Union[float, int, None]) -> List[Any]
```
Comment (guedou) [inline scapy/automaton.py:59]: Is it possible to use a stricter type than Any here?

**Review by guedou — APPROVED** [review]

File: test/contrib/automotive/ccp.uts (hunk 3 above)
Comment (gpotter2) [reply test/contrib/automotive/ccp.uts:882]: No you're right

File: scapy/automaton.py (hunk 4 above)
Comment (gpotter2) [reply scapy/automaton.py:59]: It appears python itself isn't sure how to type this better: https://github.com/python/typeshed/blob/994b69ef8f18e76689daca3947879c3d7f76173e/stdlib/select.pyi#L28

Comment (guedou) [reply scapy/automaton.py:59]: too bad =/
